gh-131591: Implement PEP 768

pablogsal opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-131591 -->
* Issue: gh-131591
<!-- /gh-issue-number -->

pablogsal committed:
gh-131591: Implement PEP 768

pablogsal committed:
Add tests and use io.open_code

pablogsal committed:
Some fixes

pablogsal committed:
Prepare for windows support

pablogsal committed:
Update windows files

ivonastojanovic committed:
Fix syntax error

pablogsal committed:
Merge pull request #97 from ivonastojanovic/fix_windows_files

Fix syntax error

godlygeek commented on Python/initconfig.c:
```diff
@@ -317,6 +318,7 @@ The following implementation-specific options are available:\n\
 -X perf: support the Linux \"perf\" profiler; also PYTHONPERFSUPPORT=1\n\
 -X perf_jit: support the Linux \"perf\" profiler with DWARF support;\n\
          also PYTHON_PERF_JIT_SUPPORT=1\n\
+-X disable-remote-debug: disable remote debugging; also PYTHON_DISABLE_REMOTE_DEBUG\n\
```
All of the others use `_` instead of `-` so this should probably be
```suggestion
-X disable_remote_debug: disable remote debugging; also PYTHON_DISABLE_REMOTE_DEBUG\n\
```

godlygeek commented on Python/initconfig.c:
```diff
@@ -1986,6 +1989,28 @@ config_init_perf_profiling(PyConfig *config)
+    if (env) {
+        active = 0;
+    }
+    const wchar_t *xoption = config_get_xoption(config, L"disable-remote-debug");
```
As above, all the other X options use underscores so this should probably be 
```suggestion
    const wchar_t *xoption = config_get_xoption(config, L"disable_remote_debug");
```

godlygeek commented on Include/cpython/pystate.h:
```diff
@@ -29,6 +29,14 @@ typedef int (*Py_tracefunc)(PyObject *, PyFrameObject *, int, PyObject *);
+# define MAX_SCRIPT_PATH_SIZE 512
+typedef struct _remote_debugger_support {
+    int enabled;
+    int debugger_pending_call;
```
We should probably be explicit about our sizes:
```suggestion
    int32_t enabled;
    int32_t debugger_pending_call;
```
Granted `int` is 32 bit on all modern platforms, but given that this goes into the offsets and needs to be read from another process with potentially different integer types (like a 64-bit process attaching to a 32-bit one or vice versa) it's better to be explicit, I think.

Or, if you'd prefer, the offsets themselves could contain the size of the field - but I think that's overkill.

godlygeek commented on Include/cpython/pystate.h:
```diff
@@ -29,6 +29,14 @@ typedef int (*Py_tracefunc)(PyObject *, PyFrameObject *, int, PyObject *);
+/* Remote debugger support */
+# define MAX_SCRIPT_PATH_SIZE 512
+typedef struct _remote_debugger_support {
+    int enabled;
```
I believe this field is unused. You added `enabled` here, but never reference it anywhere, and the offsets instead use:
```c
.remote_debugging_enabled = offsetof(PyInterpreterState, config.remote_debug),
```

I think this line should just be removed (which makes sense, anyway - we want a per-process enablement, not a per-thread enablement).

godlygeek commented on Include/internal/pycore_debug_offsets.h:
```diff
@@ -326,6 +337,14 @@ typedef struct _Py_DebugOffsets {
+    .debugger_support = { \
+        .eval_breaker = offsetof(PyThreadState, eval_breaker), \
+        .remote_debugger_support = offsetof(PyThreadState, remote_debugger_support),  \
+        .remote_debugging_enabled = offsetof(PyInterpreterState, config.remote_debug),  \
```
Should this field really go into the `.debugger_support` section? I think it should probably go into `.interpreter_state` instead, since everything else in this struct is specific to a single threadstate, except for this one field being per-interpreter.

godlygeek replied:
I'm just gonna leave it, it seems annoying to move and this was more of a nit.

godlygeek commented on Include/internal/pycore_interp_structs.h:
```diff
@@ -936,7 +936,6 @@ struct _is {
     _PyThreadStateImpl _initial_thread;
     // _initial_thread should be the last field of PyInterpreterState.
     // See https://github.com/python/cpython/issues/127117.
-
```
Accidental change?

godlygeek commented on Python/ceval_gil.c:
```diff
@@ -1319,5 +1319,70 @@ _Py_HandlePending(PyThreadState *tstate)
+
+#ifdef Py_REMOTE_DEBUG
+    const PyConfig *config = _PyInterpreterState_GetConfig(tstate->interp);
+    if (config->remote_debug) {
```
Might be worth doing:
```suggestion
    if (config->remote_debug == 1) {
```
to harden this at least a little bit against heap corruption.

godlygeek commented on Python/ceval_gil.c:
```diff
@@ -1319,5 +1319,70 @@ _Py_HandlePending(PyThreadState *tstate)
+#ifdef Py_REMOTE_DEBUG
+    const PyConfig *config = _PyInterpreterState_GetConfig(tstate->interp);
+    if (config->remote_debug) {
+        if (tstate->remote_debugger_support.debugger_pending_call) {
```
Likewise we said in the PEP that we will only be accepting 0 or 1 for `debugger_pending_call` for now.
```suggestion
        if (tstate->remote_debugger_support.debugger_pending_call == 1) {
```

godlygeek commented on Python/ceval_gil.c:
```diff
@@ -1319,5 +1319,70 @@ _Py_HandlePending(PyThreadState *tstate)
+    if (config->remote_debug) {
+        if (tstate->remote_debugger_support.debugger_pending_call) {
+            tstate->remote_debugger_support.debugger_pending_call = 0;
+            const char *path = tstate->remote_debugger_support.debugger_script_path;
```
We should strndup this, because otherwise racing with multiple remote debuggers could lead to us auditing one path and then running a different one. At least if we read it once and then reuse it, we'll be consistent even in the case of a race.

We also shouldn't trust that it's null terminated, we should check that ourselves before using it, or append a terminator ourselves.

godlygeek commented on Python/ceval_gil.c:
```diff
@@ -1319,5 +1319,70 @@ _Py_HandlePending(PyThreadState *tstate)
         }
     }
+
+#ifdef Py_REMOTE_DEBUG
```
I wonder if it's worth splitting this into a separate function, rather than putting this all in the implementation of `_Py_HandlePending`...

godlygeek commented on Python/ceval_gil.c:
```diff
@@ -1319,5 +1319,70 @@ _Py_HandlePending(PyThreadState *tstate)
+                        PyErr_FormatUnraisable("Error when converting remote debugger script path %s to wide char", path);
+                        return 0;
+                    }
+                    FILE* f = _wfopen(wpath, L"r");
```
I'm a bit confused here... On Unix, we pass `fileobj.fileno()` to `fdopen`.
On Windows, why can't we pass `fileobj.fileno()` to `_fdopen`?

I'm not seeing what we gain by reopening the file with a wide character string path/mode.

godlygeek commented on Python/ceval_gil.c:
```diff
@@ -1319,5 +1319,70 @@ _Py_HandlePending(PyThreadState *tstate)
+                    PyObject* path_obj = PyUnicode_FromString(path);
+                    if (!path_obj) {
+                        PyErr_FormatUnraisable("Error when converting remote debugger script path %s to Unicode", path);
+                        return 0;
```
Leaks `fileobj`

godlygeek commented on Python/ceval_gil.c:
```diff
@@ -1319,5 +1319,70 @@ _Py_HandlePending(PyThreadState *tstate)
+                    Py_DECREF(path_obj);
+                    if (!wpath) {
+                        PyErr_FormatUnraisable("Error when converting remote debugger script path %s to wide char", path);
+                        return 0;
```
Leaks `fileobj`

godlygeek commented on Python/ceval_gil.c:
```diff
@@ -1319,5 +1319,70 @@ _Py_HandlePending(PyThreadState *tstate)
+                    int fd = PyObject_AsFileDescriptor(fileobj);
+                    if (fd == -1) {
+                        PyErr_FormatUnraisable("Error when getting file descriptor for debugger script %s", path);
+                        return 0;
```
Leaks `fileobj`

godlygeek commented on Python/sysmodule.c:
```diff
@@ -2421,6 +2421,114 @@ sys_is_stack_trampoline_active_impl(PyObject *module)
+
+Return True if remote debugging is enabled, False otherwise.
+
+If no stack profiler is activated, this function has no effect.
```
What? This seems like a copy-paste error that should just be removed.
```suggestion
```

godlygeek commented on Python/sysmodule.c:
```diff
@@ -2421,6 +2421,114 @@ sys_is_stack_trampoline_active_impl(PyObject *module)
+to how signals are handled. There is no interface to determine when the
+code has been executed. The caller is responsible for making sure that
+the file still exists whenever the remote process tries to read it and that
+it hasn't been overwritten.
```
It's probably worth documenting that the remote process needs to be CPython, and needs to be Python 3.14+, right?

godlygeek commented on Python/sysmodule.c:
```diff
@@ -2421,6 +2421,114 @@ sys_is_stack_trampoline_active_impl(PyObject *module)
+{
+#ifdef MS_WINDOWS
+    // Get UTF-16 (wide char) version of the path for Windows
+    wchar_t *debugger_script_path_w = PyUnicode_AsWideCharString(script, NULL);
```
Above we said that `script` could be either `str|bytes`, but I believe `PyUnicode_AsWideCharString` only accepts `str`. Perhaps on Windows the `bytes` option doesn't even make sense, though. Should we document that Unix allows `bytes` but Windows does not?

godlygeek commented on Python/sysmodule.c:
```diff
@@ -2421,6 +2421,114 @@ sys_is_stack_trampoline_active_impl(PyObject *module)
+    
+    PyMem_Free(debugger_script_path_w);
+#else
+    const char *debugger_script_path = PyUnicode_AsUTF8(script);
```
Actually, looks like we're not currently handling `bytes` even for Unix.

godlygeek reviewed (commented):
Initial observations (I've looked at everything but the interesting stuff, haven't gotten to Python/remote_debugging.c or Lib/test/test_sys.py yet)

godlygeek committed:
Refactor to avoid duplicate error handling

godlygeek committed:
Check for debug offset compatibility before using the offsets

godlygeek committed:
Give a some variables shorter names

godlygeek committed:
Have read_memory/write_memory return 0 on success

Previously we were returning the number of bytes, which we didn't need
and would never be different than the argument that was passed into the
function.

godlygeek committed:
Improve an error message

ivonastojanovic committed:
Locate PyRuntime on Windows

The PyRuntime section can be found in either the main executable (python.exe) or the Python DLL (pythonXY.dll, where X and Y represent the major and minor version numbers). Scan the modules of the process to locate the PyRuntime address in one of these modules.

ivonastojanovic committed:
Read process memory on Windows

ivonastojanovic committed:
Write to process memory on Windows

ivonastojanovic committed:
Prevent attaching to a process with a different architecture

Ensure that the caller and target processes have matching architectures
before proceeding. Attaching to a 64-bit process from a 32-bit process
will fail when using CreateToolhelp32Snapshot, and attempting to attach
to a 32-bit process from a 64-bit process may cause issues during PE
file parsing. To avoid potential errors abort the operation if the
architectures differ.

godlygeek committed:
Revert "Prevent attaching to a process with a different architecture"

This reverts commit 746ecfc5b197e28c492dc5db004e81a9bf6cf48d.

That commit isn't correct. It conflates the return from `IsWow64Process`
(whether the process is running under 32-bit emulation on a 64-bit
process) with whether the process is 64-bit. A 32-bit process on 32-bit
Windows would have `IsWow64Process` set our `isWow64` flag to `FALSE`,
and we'd then incorrectly return `TRUE` from `is_process64Bit`.

godlygeek committed:
Merge pull request #98 from ivonastojanovic/external_debugger_windows_support

External debugger Windows support

godlygeek committed:
Remove unused struct member

godlygeek committed:
Set an explicit size for a struct member

This is useful because debuggers will write to this variable remotely,
and it's helpful for it to have a well known size rather than one that
could vary per platform.

godlygeek committed:
Revert an accidental whitespace change

godlygeek committed:
Factor running a debugger script into a helper function

godlygeek committed:
Harden remote debugging integration

Require the flags for turning this on to be set to exactly 1 to avoid
accidentally triggering remote debugging in the case of heap corruption.
Make a heap copy of the script path before using it to avoid the buffer
being overwritten while we're still using it by another debugger.

godlygeek committed:
Use _fdopen to open the debugger script on Windows

This avoids the need to reopen it by (wide character) path.

godlygeek committed:
Ensure the debugger script is always closed

Previously it was leaked if `PyObject_AsFileDescriptor` failed.

godlygeek committed:
Fix incorrect docstring

I think this was a copy paste error.

godlygeek committed:
Document restrictions for remote_exec

The remote process must be a compatible version.

Explain our compatibility rules.

godlygeek committed:
Simplify handling of Windows paths

We can clean this up by freeing strings as soon as we're done with them,
and remove some duplicate code.

godlygeek committed:
Make remote_exec accept `bytes` paths

This was documented as working, but in fact we assumed that the path
would always be a unicode string. Factor the version that only accepts a
unicode string into a helper function, and have the caller call
`os.fsdecode` on the input path, which accepts either a byte string or a
unicode string and always returns a unicode string.

godlygeek committed:
Remove a debugging printf

godlygeek committed:
Fix a remote read to use the correct type

This field in the structure is defined as an `unsigned long`, not as
a `pid_t`.

godlygeek committed:
Improve the error message when we can't find the requested thread

godlygeek committed:
Remove a useless check

We already did a `strlen` on the `debugger_script_path`, so it's too late
to check if it's NULL here!

godlygeek committed:
Only accept a flag of 1 to mean remote debugging is enabled

This helps to harden a bit against heap corruption.

graingert commented on Lib/test/test_sys.py:
```diff
@@ -1923,5 +1927,205 @@ def write(self, s):
+        self.addCleanup(os_helper.unlink, target)
+
+        # Find an available port for the socket
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
```
test.support.socket_helper.find_unused_port ?

could you use a socketpair and pass the fd to the child `Popen(pass_fds=[sock.fileno()])`?

pablogsal replied:
Done

pablogsal replied:
Ah, I am reverting this because this is not supported on Windows:

```
======================================================================
FAIL: test_remote_exec_with_exception (test.test_sys.TestRemoteExec.test_remote_exec_with_exception)
Test remote exec with an exception raised in the target process
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\cpython\cpython\Lib\test\test_sys.py", line 2107, in test_remote_exec_with_exception
    returncode, stdout, stderr = self._run_remote_exec_test(script)
                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
  File "D:\a\cpython\cpython\Lib\test\test_sys.py", line 2020, in _run_remote_exec_test
    with subprocess.Popen(cmd,
         ~~~~~~~~~~~~~~~~^^^^^
                          stdout=subprocess.PIPE,
                          ^^^^^^^^^^^^^^^^^^^^^^^
    ...<2 lines>...
                          pass_fds=[child_sock.fileno()],
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                          ) as proc:
                          ^
  File "D:\a\cpython\cpython\Lib\subprocess.py", line 1035, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                        pass_fds, cwd, env,
                        ^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
                        gid, gids, uid, umask,
                        ^^^^^^^^^^^^^^^^^^^^^^
                        start_new_session, process_group)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\a\cpython\cpython\Lib\subprocess.py", line 1456, in _execute_child
    assert not pass_fds, "pass_fds not supported on Windows."
           ^^^^^^^^^^^^
AssertionError: pass_fds not supported on Windows.
```

graingert commented on Lib/test/test_sys.py:
```diff
@@ -1923,5 +1927,205 @@ def write(self, s):
+                              env=env) as proc:
+            try:
+                # Accept connection from target process
+                client_socket, _ = server_socket.accept()
```
you can close the server socket as soon as the client_socket is accepted

pablogsal replied:
Done

pablogsal committed:
Merge remote-tracking branch 'upstream/main'

pablogsal committed:
Lint

pablogsal committed:
simplify socket handling

pablogsal committed:
Add NEWS entry

pablogsal committed:
Add more docs
